### ISSUE TITLE
SAVE GAME then LOAD GAME on TRAPPED PLAY GAME STANDALONE disables the weapon so cannot pick up

### ISSUE THREAD

**Issue by LeeBamberTGC** [TheGameCreators/GameGuruRepo#2642]
Run the TRAPPED demo as a STANDALONE, notice after loading a saved game position, the gun cannot be collected in the first room

**Comment by ZakJudges** [comment]
@LeeBamberTGC This seems to be a bigger issue than I thought. As well as not being able to collect the gun, you also cannot open doors. I took a closer look and noticed that the LUA side of things was taking a LONG time when the game has been loaded vs started from the beginning:

![Screenshot 2022-03-28 102216](https://user-images.githubusercontent.com/32545391/160368615-0846c516-2b5c-4efe-8075-48f5abab4986.png)

The offending code is in lua_loop_begin, under this comment: "// 020316 - call the global loop once per cycle (for things like loading game states)"

When LuaExectute(GlobalLoop) is called, it returns a different value when the game is loaded vs started from beginning. But I couldn't see any errors, so I couldn't figure out why that is the case. Any ideas of where to look? Thanks :)

**Comment by LeeBamberTGC** [comment]
@ZakJudges The fix is now in the file "Max\Files\titlesbank\gamedata.lua" and relates to a crash when loading a saved game position that used a behavior script that needed to save three dimensional arrays and crashed the script. After that, a successful flag could not be set to one and so caused all logic to stay disabled as the load was not successful. Have a look at this cascade so you know what I did and ask questions if you like :)

**Comment by ZakJudges** [comment]
@LeeBamberTGC Thanks for the info :) I understand how the crash was prevented, though I'm not sure if I would be able to fix it if something similar came up in the future. Any chance you could share your thought process, and what led you to your fix? Specifically, how did you find the error was in gamedata.lua, and how did you find the cause of the crash? Is there a debugger for lua I can use? Thanks :)

**Comment by LeeBamberTGC** [comment]
@ZakJudges Well your own investigation was a great start as you noticed where the two pathways (good and bad) diverged. From there I looked at why GlobalLoop was taking so long, and on looking closer found that the after the LUA begin function you mentioned, the actual loop that processes individual scripts was being skipped as a flag that should be zero was set to 2.  This lead me to asking why this was not set to zero, which in turn lead me to only two places where this value was set, and one of those suggested it was because the script did not return 'succesful=1' which would inform the rest of the code that the load was successful. Without this, the aforementioned flag would always stay at 2. I then added some debug code to see how far the load function (nested down into gamedata.lua) got before crashing out, and it turned out that the old code could not execute a string that had a three dimensional array, it would actually crash.  I then googled alternative ways to get a LUA string and convert it to a function that could be called, and the PCALL method ensured that if it was an illegal call, it could be safely ignored. The line that could not be called is now skipped, but it relates to globals that are not needed when saving game positions and only temporary variables for the object in question. Bottom line, a better way to load in the LUA globals (see "Documents\GameGuruApps\GameGuruMAX\Files\savegames\gameslot1.dat") was the solution, which meant no more crashing the script, a successful load, a flag that correctly set itself to zero and the game could perform all the scripts for all of the objects, and ultimately the player could pick up the ammo and gun after a load game event. Phew!
